server/src/com/thoughtworks/go/config/CachedGoConfig.java

```diff
@@ -18,36 +18,207 @@
+import com.thoughtworks.go.serverhealth.ServerHealthState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.thoughtworks.go.server.service.GoConfigService.INVALID_CRUISE_CONFIG_XML;
+import static com.thoughtworks.go.util.ExceptionUtils.bomb;
+
+/**
+ * @understands when to reload the config file or other config source
+ */
+@Component
+public class CachedGoConfig {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CachedGoConfig.class);
+    private final GoFileConfigDataSource dataSource;
+    private final ServerHealthService serverHealthService;
+    private List<ConfigChangedListener> listeners = new ArrayList<>();
+    private volatile CruiseConfig currentConfig;
+    private volatile CruiseConfig currentConfigForEdit;
+    private volatile GoConfigHolder configHolder;
+    private volatile Exception lastException;
+
+    @Autowired
+    public CachedGoConfig(ServerHealthService serverHealthService,
+                          GoFileConfigDataSource dataSource) {
+        this.serverHealthService = serverHealthService;
+        this.dataSource = dataSource;
+    }
+
+    public static List<ConfigErrors> validate(CruiseConfig config) {
+        List<ConfigErrors> validationErrors = new ArrayList<ConfigErrors>();
+        validationErrors.addAll(config.validateAfterPreprocess());
+        return validationErrors;
+    }
+
+    //used in tests
+    public void throwExceptionIfExists() {
+        if (lastException != null) {
+            throw bomb("Invalid config file", lastException);
+        }
+    }
+
+    //NOTE: This method is called on a thread from Spring
+    public void onTimer() {
+        this.forceReload();
+    }
+
+    public void forceReload() {
+        LOGGER.debug("Config file (on disk) update check is in queue");
+        synchronized (GoConfigWriteLock.class) {
+            LOGGER.debug("Config file (on disk) update check is in progress");
+            try {
+                GoConfigHolder configHolder = dataSource.load();
+                if (configHolder != null) {
+                    saveValidConfigToCacheAndNotifyConfigChangeListeners(configHolder);
+                }
+            } catch (Exception e) {
+                LOGGER.warn("Error loading cruise-config.xml from disk, keeping previous one", e);
+                saveConfigError(e);
+            }
+        }
+    }
+
+    private synchronized void saveConfigError(Exception e) {
+        this.lastException = e;
+        ServerHealthState state = ServerHealthState.error(INVALID_CRUISE_CONFIG_XML, GoConfigValidity.invalid(e).errorMessage(), HealthStateType.invalidConfig());
+        serverHealthService.update(state);
+    }
+
+    public CruiseConfig loadForEditing() {
+        return currentConfigForEdit;
```

Comment:
I guess there's never a time that this is null. But, we do sometimes check for null (like in `currentConfig` below). Why not here? I know `loadConfigIfNull` called from ApplicationInitializer, but if so, why check in `currentConfig`?